whylog/constraints/__init__.py

```diff
@@ -2,42 +2,189 @@
+
+        For Front to ask user for proper count of groups.
+        2, None - at least 2 groups
+        2, 2 - exactly 2 groups
+        """
+        return cls.MIN_GROUPS_COUNT, cls.MAX_GROUPS_COUNT
+
+    @classmethod
+    def verify(cls, group_contents, param_dict):
+        """
+        Verifies constraint for given params in param_dict and groups contents.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: list of groups contents,
+
+        For LogReader and Teacher verification.
+        It must be optimized as well as possible (for LogReader).
+        """
+
+        raise NotImplementedError("Subclass should implement this")
 
 
 class TimeConstraint(AbstractConstraint):
-    def __init__(self, line_earlier, line_later, min_delta=None, max_delta=None):
-        pass
+    """
+    Time delta between two dates must be greater than 'max_delta' and lower than 'min_delta'
+    """
+
+    TYPE = ConstraintType.TIME_DELTA
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = 2
+
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    PARAMS = [MIN_DELTA, MAX_DELTA]
+
+    def __init__(self, groups, param_dict):
+        """
+        :param groups: First element of groups is a group with earlier date.
+
+        I.e:
+        TimeConstraint(
+            [(0, 1), (2, 1)]
+            {'min_delta': datetime.timedelta(seconds=1), 'max_delta': datetime.timedelta(seconds=10)},
+        )
+        """
+        super(TimeConstraint, self).__init__(groups, param_dict)
 
+    def _check_constructor_params(self):
+        param_names = set(self.params.keys())
+        correct_param_names = set(self.get_param_names())
+        if (self.MIN_DELTA not in param_names and self.MAX_DELTA not in param_names)\
+                or param_names - correct_param_names:
+            raise ConstructorParamsError(self.TYPE, correct_param_names, param_names)
 
-class IdenticalIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    @classmethod
+    def verify(cls, group_contents, param_dict):
+        #TODO
+        raise NotImplementedError
 
 
-class AnyValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+class IdenticalConstraint(AbstractConstraint):
+    """
+    Contents of groups must be identical.
+    """
 
+    TYPE = ConstraintType.IDENTICAL
 
-class DifferentValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = None
 
+    PARAMS = []
 
-class ConstIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    def __init__(self, groups, param_dict=None):
+        """
+        I.e:
+        IdenticalConstraint(
+            [(1, 2), (2, 4)]
+        )
+        """
+        super(IdenticalConstraint, self).__init__(groups, param_dict)
 
+    def _check_constructor_params(self):
```

Comment:
This can be removed now that we have a generic verification method (described in another comment)